File: quick_sort.rb
# define algorithm here (you'll need lo and hi for the beginning/endpoints on the recursive call)
def quicksort(arr, lo, hi)
  if lo < hi
    new_pivot_index = partition(arr, lo, hi)
    quicksort(arr, lo, new_pivot_index - 1)
    quicksort(arr, new_pivot_index + 1, hi)
  end

  return arr
end

# partition function (selects a pivot, sorts into partitions, and returns the array index of the pivot)
def partition(arr, lo, hi)
  pivot = arr[hi]
  left = lo

  for element in (lo...hi)
    if arr[element] <= pivot
      arr[left], arr[element] = arr[element], arr[left]
      left += 1
    end
  end

  arr[left], arr[hi] = arr[hi], arr[left]
  return left
end


# testing quicksort
test = (1..10).to_a.shuffle
quicksort(test, 0, test.length-1)

if test == (1..10).to_a
  puts 'The sort worked!'
else
  puts 'Noooo, the sort failed!'
end

File: quickSort.js
// define algorithm here (you'll need lo and hi for the beginning/endpoints on the recursive call)
function quicksort(arr, lo, hi) {
  if (lo < hi) {
    var pivotIndex = partition(arr, lo, hi);
    quicksort(arr, lo, pivotIndex - 1);
    quicksort(arr, pivotIndex + 1, hi);
  }
  return arr;
}

// partition function (selects a pivot, sorts into partitions, and returns the array index of the pivot)
function partition(arr, lo, hi) {
  var pivotValue = arr[hi];
  var left = lo;
  var temp;

  for (let i = lo; i < hi; i++) {
    if (arr[i] <= pivotValue) {
      // arr[left], arr[i] = arr[i], arr[left]
      temp = arr[left];
      arr[left] = arr[i];
      arr[i] = temp;
      left++;
    }
  }

  // arr[left], arr[hi] = arr[hi], arr[left]
  temp = arr[left];
  arr[left] = arr[hi];
  arr[hi] = temp;
  return left;
}

// testing quicksort
var a = [4, 92, 1, 39, 19, 93, 49, 10, 99, 103, 13];
quicksort(a, 0, a.length-1)

console.log(a);
